File: avm-transpiler/src/opcodes.rs
/// All AVM opcodes
/// Keep updated with TS, cpp, and docs protocol specs!
#[allow(clippy::upper_case_acronyms, dead_code, non_camel_case_types)]
#[derive(PartialEq, Copy, Clone, Debug, Eq, Hash)]
pub enum AvmOpcode {
    // Compute
    ADD_8,
    ADD_16,
    SUB_8,
    SUB_16,
    MUL_8,
    MUL_16,
    DIV_8,
    DIV_16,
    FDIV_8,
    FDIV_16,
    EQ_8,
    EQ_16,
    LT_8,
    LT_16,
    LTE_8,
    LTE_16,
    AND_8,
    AND_16,
    OR_8,
    OR_16,
    XOR_8,
    XOR_16,
    NOT_8,
    NOT_16,
    SHL_8,
    SHL_16,
    SHR_8,
    SHR_16,
    CAST_8,
    CAST_16,
    // Execution environment
    GETENVVAR_16,
    CALLDATACOPY,
    RETURNDATASIZE,
    RETURNDATACOPY,
    // Control flow
    JUMP_32,
    JUMPI_32,
    INTERNALCALL,
    INTERNALRETURN,
    // Memory
    SET_8,
    SET_16,
    SET_32,
    SET_64,
    SET_128,
    SET_FF,
    MOV_8,
    MOV_16,
    // World state
    SLOAD,
    SSTORE,
    NOTEHASHEXISTS,
    EMITNOTEHASH,
    NULLIFIEREXISTS,
    EMITNULLIFIER,
    L1TOL2MSGEXISTS,
    GETCONTRACTINSTANCE,
    EMITUNENCRYPTEDLOG,
    SENDL2TOL1MSG,
    // External calls
    CALL,
    STATICCALL,
    RETURN,
    REVERT_8,
    REVERT_16,
    // Misc
    DEBUGLOG,
    // Gadgets
    POSEIDON2,
    SHA256COMPRESSION,
    KECCAKF1600,
    ECADD,
    MSM,
    // Conversions
    TORADIXBE,
}

impl AvmOpcode {
    pub fn name(&self) -> &'static str {
        match self {
            // Compute
            // Compute - Arithmetic
            AvmOpcode::ADD_8 => "ADD_8",
            AvmOpcode::ADD_16 => "ADD_16",
            AvmOpcode::SUB_8 => "SUB_8",
            AvmOpcode::SUB_16 => "SUB_16",
            AvmOpcode::MUL_8 => "MUL_8",
            AvmOpcode::MUL_16 => "MUL_16",
            AvmOpcode::DIV_8 => "DIV_8",
            AvmOpcode::DIV_16 => "DIV_16",
            AvmOpcode::FDIV_8 => "FDIV_8",
            AvmOpcode::FDIV_16 => "FDIV_16",
            // Compute - Comparators
            AvmOpcode::EQ_8 => "EQ_8",
            AvmOpcode::EQ_16 => "EQ_16",
            AvmOpcode::LT_8 => "LT_8",
            AvmOpcode::LT_16 => "LT_16",
            AvmOpcode::LTE_8 => "LTE_8",
            AvmOpcode::LTE_16 => "LTE_16",
            // Compute - Bitwise
            AvmOpcode::AND_8 => "AND_8",
            AvmOpcode::AND_16 => "AND_16",
            AvmOpcode::OR_8 => "OR_8",
            AvmOpcode::OR_16 => "OR_16",
            AvmOpcode::XOR_8 => "XOR_8",
            AvmOpcode::XOR_16 => "XOR_16",
            AvmOpcode::NOT_8 => "NOT_8",
            AvmOpcode::NOT_16 => "NOT_16",
            AvmOpcode::SHL_8 => "SHL_8",
            AvmOpcode::SHL_16 => "SHL_16",
            AvmOpcode::SHR_8 => "SHR_8",
            AvmOpcode::SHR_16 => "SHR_16",
            // Compute - Type Conversions
            AvmOpcode::CAST_8 => "CAST_8",
            AvmOpcode::CAST_16 => "CAST_16",

            // Execution Environment
            AvmOpcode::GETENVVAR_16 => "GETENVVAR_16",
            // Execution Environment - Calldata
            AvmOpcode::CALLDATACOPY => "CALLDATACOPY",
            AvmOpcode::RETURNDATASIZE => "RETURNDATASIZE",
            AvmOpcode::RETURNDATACOPY => "RETURNDATACOPY",

            // Machine State
            // Machine State - Internal Control Flow
            AvmOpcode::JUMP_32 => "JUMP_32",
            AvmOpcode::JUMPI_32 => "JUMPI_32",
            AvmOpcode::INTERNALCALL => "INTERNALCALL",
            AvmOpcode::INTERNALRETURN => "INTERNALRETURN",
            // Machine State - Memory
            AvmOpcode::SET_8 => "SET_8",
            AvmOpcode::SET_16 => "SET_16",
            AvmOpcode::SET_32 => "SET_32",
            AvmOpcode::SET_64 => "SET_64",
            AvmOpcode::SET_128 => "SET_128",
            AvmOpcode::SET_FF => "SET_FF",
            AvmOpcode::MOV_8 => "MOV_8",
            AvmOpcode::MOV_16 => "MOV_16",

            // World State
            AvmOpcode::SLOAD => "SLOAD",   // Public Storage
            AvmOpcode::SSTORE => "SSTORE", // Public Storage
            AvmOpcode::NOTEHASHEXISTS => "NOTEHASHEXISTS", // Notes & Nullifiers
            AvmOpcode::EMITNOTEHASH => "EMITNOTEHASH", // Notes & Nullifiers
            AvmOpcode::NULLIFIEREXISTS => "NULLIFIEREXISTS", // Notes & Nullifiers
            AvmOpcode::EMITNULLIFIER => "EMITNULLIFIER", // Notes & Nullifiers
            AvmOpcode::L1TOL2MSGEXISTS => "L1TOL2MSGEXISTS", // Messages

            // Accrued Substate
            AvmOpcode::EMITUNENCRYPTEDLOG => "EMITUNENCRYPTEDLOG",
            AvmOpcode::SENDL2TOL1MSG => "SENDL2TOL1MSG",
            AvmOpcode::GETCONTRACTINSTANCE => "GETCONTRACTINSTANCE",

            // Control Flow - Contract Calls
            AvmOpcode::CALL => "CALL",
            AvmOpcode::STATICCALL => "STATICCALL",
            AvmOpcode::RETURN => "RETURN",
            AvmOpcode::REVERT_8 => "REVERT_8",
            AvmOpcode::REVERT_16 => "REVERT_16",

            // Misc
            AvmOpcode::DEBUGLOG => "DEBUGLOG",

            // Gadgets
            AvmOpcode::POSEIDON2 => "POSEIDON2",
            AvmOpcode::SHA256COMPRESSION => "SHA256COMPRESSION",
            AvmOpcode::KECCAKF1600 => "KECCAKF1600",
            AvmOpcode::ECADD => "ECADD",
            AvmOpcode::MSM => "MSM",
            // Conversions
            AvmOpcode::TORADIXBE => "TORADIXBE",
        }
    }
}

File: barretenberg/cpp/src/barretenberg/vm2/common/opcodes.cpp
#include "barretenberg/vm2/common/opcodes.hpp"

#include <ostream>
#include <string>

namespace bb::avm2 {

std::ostream& operator<<(std::ostream& os, const WireOpCode& op)
{
    switch (op) {
    case WireOpCode::ADD_8:
        os << "ADD_8";
        break;
    case WireOpCode::ADD_16:
        os << "ADD_16";
        break;
    case WireOpCode::SUB_8:
        os << "SUB_8";
        break;
    case WireOpCode::SUB_16:
        os << "SUB_16";
        break;
    case WireOpCode::MUL_8:
        os << "MUL_8";
        break;
    case WireOpCode::MUL_16:
        os << "MUL_16";
        break;
    case WireOpCode::DIV_8:
        os << "DIV_8";
        break;
    case WireOpCode::DIV_16:
        os << "DIV_16";
        break;
    case WireOpCode::FDIV_8:
        os << "FDIV_8";
        break;
    case WireOpCode::FDIV_16:
        os << "FDIV_16";
        break;
    case WireOpCode::EQ_8:
        os << "EQ_8";
        break;
    case WireOpCode::EQ_16:
        os << "EQ_16";
        break;
    case WireOpCode::LT_8:
        os << "LT_8";
        break;
    case WireOpCode::LT_16:
        os << "LT_16";
        break;
    case WireOpCode::LTE_8:
        os << "LTE_8";
        break;
    case WireOpCode::LTE_16:
        os << "LTE_16";
        break;
    case WireOpCode::AND_8:
        os << "AND_8";
        break;
    case WireOpCode::AND_16:
        os << "AND_16";
        break;
    case WireOpCode::OR_8:
        os << "OR_8";
        break;
    case WireOpCode::OR_16:
        os << "OR_16";
        break;
    case WireOpCode::XOR_8:
        os << "XOR_8";
        break;
    case WireOpCode::XOR_16:
        os << "XOR_16";
        break;
    case WireOpCode::NOT_8:
        os << "NOT_8";
        break;
    case WireOpCode::NOT_16:
        os << "NOT_16";
        break;
    case WireOpCode::SHL_8:
        os << "SHL_8";
        break;
    case WireOpCode::SHL_16:
        os << "SHL_16";
        break;
    case WireOpCode::SHR_8:
        os << "SHR_8";
        break;
    case WireOpCode::SHR_16:
        os << "SHR_16";
        break;
    case WireOpCode::CAST_8:
        os << "CAST_8";
        break;
    case WireOpCode::CAST_16:
        os << "CAST_16";
        break;
    case WireOpCode::GETENVVAR_16:
        os << "GETENVVAR_16";
        break;
    case WireOpCode::CALLDATACOPY:
        os << "CALLDATACOPY";
        break;
    case WireOpCode::RETURNDATASIZE:
        os << "RETURNDATASIZE";
        break;
    case WireOpCode::RETURNDATACOPY:
        os << "RETURNDATACOPY";
        break;
    case WireOpCode::JUMP_32:
        os << "JUMP_32";
        break;
    case WireOpCode::JUMPI_32:
        os << "JUMPI_32";
        break;
    case WireOpCode::INTERNALCALL:
        os << "INTERNALCALL";
        break;
    case WireOpCode::INTERNALRETURN:
        os << "INTERNALRETURN";
        break;
    case WireOpCode::SET_8:
        os << "SET_8";
        break;
    case WireOpCode::SET_16:
        os << "SET_16";
        break;
    case WireOpCode::SET_32:
        os << "SET_32";
        break;
    case WireOpCode::SET_64:
        os << "SET_64";
        break;
    case WireOpCode::SET_128:
        os << "SET_128";
        break;
    case WireOpCode::SET_FF:
        os << "SET_FF";
        break;
    case WireOpCode::MOV_8:
        os << "MOV_8";
        break;
    case WireOpCode::MOV_16:
        os << "MOV_16";
        break;
    case WireOpCode::SLOAD:
        os << "SLOAD";
        break;
    case WireOpCode::SSTORE:
        os << "SSTORE";
        break;
    case WireOpCode::NOTEHASHEXISTS:
        os << "NOTEHASHEXISTS";
        break;
    case WireOpCode::EMITNOTEHASH:
        os << "EMITNOTEHASH";
        break;
    case WireOpCode::NULLIFIEREXISTS:
        os << "NULLIFIEREXISTS";
        break;
    case WireOpCode::EMITNULLIFIER:
        os << "EMITNULLIFIER";
        break;
    case WireOpCode::L1TOL2MSGEXISTS:
        os << "L1TOL2MSGEXISTS";
        break;
    case WireOpCode::GETCONTRACTINSTANCE:
        os << "GETCONTRACTINSTANCE";
        break;
    case WireOpCode::EMITUNENCRYPTEDLOG:
        os << "EMITUNENCRYPTEDLOG";
        break;
    case WireOpCode::SENDL2TOL1MSG:
        os << "SENDL2TOL1MSG";
        break;
    case WireOpCode::CALL:
        os << "CALL";
        break;
    case WireOpCode::STATICCALL:
        os << "STATICCALL";
        break;
    case WireOpCode::RETURN:
        os << "RETURN";
        break;
    case WireOpCode::REVERT_8:
        os << "REVERT_8";
        break;
    case WireOpCode::REVERT_16:
        os << "REVERT_16";
        break;
    case WireOpCode::DEBUGLOG:
        os << "DEBUGLOG";
        break;
    case WireOpCode::POSEIDON2PERM:
        os << "POSEIDON2PERM";
        break;
    case WireOpCode::SHA256COMPRESSION:
        os << "SHA256COMPRESSION";
        break;
    case WireOpCode::KECCAKF1600:
        os << "KECCAKF1600";
        break;
    case WireOpCode::ECADD:
        os << "ECADD";
        break;
    case WireOpCode::MSM:
        os << "MSM";
        break;
    case WireOpCode::TORADIXBE:
        os << "TORADIXBE";
        break;
    case WireOpCode::LAST_OPCODE_SENTINEL:
        os << "LAST_OPCODE_SENTINEL";
        break;
    }
    return os;
}

std::ostream& operator<<(std::ostream& os, const ExecutionOpCode& op)
{
    switch (op) {
    case ExecutionOpCode::ADD:
        os << "ADD";
        break;
    case ExecutionOpCode::SET:
        os << "SET";
        break;
    case ExecutionOpCode::MOV:
        os << "MOV";
        break;
    case ExecutionOpCode::CALL:
        os << "CALL";
        break;
    case ExecutionOpCode::JUMP:
        os << "JUMP";
        break;
    case ExecutionOpCode::JUMPI:
        os << "JUMPI";
        break;
    case ExecutionOpCode::RETURN:
        os << "RETURN";
        break;
    }
    return os;
}

} // namespace bb::avm2